r/src/array_to_vector.cpp

```diff
@@ -1275,7 +1276,22 @@ cpp11::writable::list to_dataframe_parallel(
 
 // [[arrow::export]]
 SEXP Array__as_vector(const std::shared_ptr<arrow::Array>& array) {
-  return arrow::r::ArrayVector__as_vector(array->length(), array->type(), {array});
+  auto type = array->type();
+
+#if defined(HAS_ALTREP)
+  if (array->null_count() == 0 && arrow::r::GetBoolOption("arrow.use_altrep", true)) {
+    switch (type->id()) {
+      case arrow::Type::DOUBLE:
+        return arrow::r::Make_array_nonull_dbl_vector(array);
+      case arrow::Type::INT32:
+        return arrow::r::Make_array_nonull_int_vector(array);
+      default:
```

Comment:
Is there a plan to support other types here?

Reply:
These are the only cases where we can use this mechanism. Although we could also use it for int64 but at the expense of automatic down casting to int32. 

We might however implement other things on top of altrep, e.g. strings or vectors with nulls but this will have to be implemented differently.